src/source-dpdk.c

```diff
@@ -0,0 +1,525 @@
+        rte_pktmbuf_free(mbuf_array[i]);
+    }
+}
+
+static uint64_t CyclesToMicroseconds(const uint64_t cycles)
+{
+    const uint64_t ticks_per_us = rte_get_tsc_hz() / 1000000;
+    return cycles / ticks_per_us;
+}
+
+static uint64_t CyclesToSeconds(const uint64_t cycles)
+{
+    const uint64_t ticks_per_s = rte_get_tsc_hz();
+    return cycles / ticks_per_s;
+}
+
+static void CyclesAddToTimeval(
+        const uint64_t cycles, struct timeval *orig_tv, struct timeval *new_tv)
+{
+    uint64_t usec = CyclesToMicroseconds(cycles) + orig_tv->tv_usec;
+    new_tv->tv_sec = orig_tv->tv_sec + usec / 1000000;
+    new_tv->tv_usec = (usec % 1000000);
+}
+
+static void DPDKSetTimevalOfMachineStart(struct timeval *tv)
+{
+    gettimeofday(tv, NULL);
+    tv->tv_sec -= DPDKGetSeconds();
+}
+
+/**
+ * Initializes real_tv to the correct real time. Adds TSC counter value to the timeval of
+ * the machine start
+ * @param machine_start_tv - timestamp when the machine was started
+ * @param real_tv
+ */
+static void DPDKSetTimevalReal(struct timeval *machine_start_tv, struct timeval *real_tv)
+{
+    CyclesAddToTimeval(rte_get_tsc_cycles(), machine_start_tv, real_tv);
+}
+
+/* get number of seconds from the reset of TSC counter (typically from the machine start) */
+static uint64_t DPDKGetSeconds()
+{
+    return CyclesToSeconds(rte_get_tsc_cycles());
+}
+
+/**
+ * \brief Registration Function for RecieveDPDK.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleReceiveDPDKRegister(void)
+{
+    tmm_modules[TMM_RECEIVEDPDK].name = "ReceiveDPDK";
+    tmm_modules[TMM_RECEIVEDPDK].ThreadInit = ReceiveDPDKThreadInit;
+    tmm_modules[TMM_RECEIVEDPDK].Func = NULL;
+    tmm_modules[TMM_RECEIVEDPDK].PktAcqLoop = ReceiveDPDKLoop;
+    tmm_modules[TMM_RECEIVEDPDK].PktAcqBreakLoop = NULL;
+    tmm_modules[TMM_RECEIVEDPDK].ThreadExitPrintStats = ReceiveDPDKThreadExitStats;
+    tmm_modules[TMM_RECEIVEDPDK].ThreadDeinit = ReceiveDPDKThreadDeinit;
+    tmm_modules[TMM_RECEIVEDPDK].cap_flags = SC_CAP_NET_RAW;
+    tmm_modules[TMM_RECEIVEDPDK].flags = TM_FLAG_RECEIVE_TM;
+}
+
+/**
+ * \brief Registration Function for DecodeDPDK.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleDecodeDPDKRegister(void)
+{
+    tmm_modules[TMM_DECODEDPDK].name = "DecodeDPDK";
+    tmm_modules[TMM_DECODEDPDK].ThreadInit = DecodeDPDKThreadInit;
+    tmm_modules[TMM_DECODEDPDK].Func = DecodeDPDK;
+    tmm_modules[TMM_DECODEDPDK].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_DECODEDPDK].ThreadDeinit = DecodeDPDKThreadDeinit;
+    tmm_modules[TMM_DECODEDPDK].cap_flags = 0;
+    tmm_modules[TMM_DECODEDPDK].flags = TM_FLAG_DECODE_TM;
+}
+
+static inline void DPDKDumpCounters(DPDKThreadVars *ptv)
+{
+#ifdef PACKET_STATISTICS
+    // Prevents access to other fields of stats structure
+    if (unlikely(ptv->queue_id >= RTE_ETHDEV_QUEUE_STAT_CNTRS))
+        return;
+
+    struct rte_eth_stats eth_stats;
+    rte_eth_stats_get(ptv->port_id, &eth_stats);
```

Comment:
This function returns a value ... should not proceed if an error occurs.

Reply:
I have added a check to validate the return value. I return from the function in case of an error.